File: src/main/java/com/pommedeterresautee/twoborange3/BusyBox.java
package com.pommedeterresautee.twoborange3;

import android.content.Context;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;


public class BusyBox {

    private static final String BUSYBOX_FILENAME = "busybox";
    private static final String BUSYBOX_FILENAME_INTERN = "busybox-linus";
    private static final long BUSYBOX_SIZE = 1165484l;

    public static void startCopyAsyncBusyboxIfNeeded(final Context context) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                copyBusyBoxIfNotYetDone(context.getApplicationContext());
            }
        }).run();
    }

    /**
     * Get the path to the BusyBox file.
     *
     * @param context
     * @return path
     */
    public static String getBusyBoxPath(Context context) {
//        if (RootPref.isBusyboxInternal()) {
//            return getInternalBusybox(context).getAbsolutePath();
//        } else {
//            return BUSYBOX_FILENAME;
//        }
        return "";
    }

    private static File getInternalBusybox(Context context) {
        return new File(context.getFilesDir(), BUSYBOX_FILENAME);
    }

    private static boolean copyBusyBoxIfNotYetDone(Context context) {
        Context appContext = context.getApplicationContext();
        if (!checkBusyBox(appContext)) {
            return copyBusyBox(appContext) > 0l;
        } else {
            return false;
        }
    }

    /**
     * Copy BusyBox executable to the private data folder.
     *
     * @param context
     */
    private static long copyBusyBox(Context context) {

        long sizeCopied = 0l;
        try {
            File busybox = getInternalBusybox(context);
            if (busybox.exists()) {
//                Preconditions.checkState(busybox.delete(), "Impossible to delete existing busybox");
            }
            FileOutputStream fos = context.openFileOutput(BUSYBOX_FILENAME, Context.MODE_PRIVATE);
            InputStream is = context.getAssets().open(BUSYBOX_FILENAME_INTERN);
//            sizeCopied = ByteStreams.copy(is, fos);
//            Preconditions.checkArgument(sizeCopied > 0l, "Failed to copy busybox to this destination: " + busybox.getAbsolutePath());
            boolean readable = busybox.setReadable(false);
            boolean writable = busybox.setWritable(false);
            boolean exec = busybox.setExecutable(true, true);
//            Preconditions.checkArgument(readable && writable && exec, "Failed to change properties of this file: " + busybox.getAbsolutePath());
            return sizeCopied;
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return sizeCopied;
    }


    /**
     * Check if BusyBox application has been copied
     * in the private data folder of the application.
     *
     * @param context
     * @return true if already copied.
     */
    private static boolean checkBusyBox(Context context) {
        File busybox = getInternalBusybox(context);
        return busybox.exists()
                && busybox.length() == BUSYBOX_SIZE
                && busybox.canExecute();
    }
}

File: src/main/scala/com/pommedeterresautee/twoborange3/Common/Busybox.scala
package com.pommedeterresautee.twoborange3.Common

import java.io._
import android.content.Context
import rx.lang.scala._
import rx.lang.scala.concurrency.Schedulers
import rx.lang.scala.Notification.{OnError, OnCompleted, OnNext}
import org.scaloid.common._
import com.pommedeterresautee.twoborange3.R


object Busybox {
  private val BUSYBOX_FILENAME_EXTERN: String = "busybox"
  private val BUSYBOX_FILENAME_INTERN: String = "busybox-linus"
  private val BUSYBOX_SIZE: Long = 1165484l

  /**
   * Proceed to the installation of Busybox in a thread safe way.
   * @param context Activity one
   * @return a Rx Unsubscribe object
   */
  def startCopyAsyncBusyboxIfNeeded(implicit context: Context) = {
    val unsubsb = Observable{
      copyBusyBoxIfNotYetDone(context.getApplicationContext)
    }
      .subscribeOn(Schedulers.newThread)
      .observeOn(Schedulers.currentThread)
      .materialize
      .subscribe(n => n match {
      case OnNext(v) if v => longToast(R.string.busybox_installed)
      case OnCompleted(c) => //Never called
      case OnError(err) => err match {case err: Exception => longToast(context.getString(R.string.busybox_error_installation, err.getMessage))}
    })
    Some(unsubsb)
  }

  /**
   * Get the path to the BusyBox file.
   *
   * @param context
   * @return path
   */
  def getBusyBoxPath(context: Context): String = {
    if (true/*RootPref.isBusyboxInternal*/) {
      getInternalBusybox(context).getAbsolutePath
    }
    else {
      BUSYBOX_FILENAME_EXTERN
    }
  }

  private def getInternalBusybox(context: Context): File = {
    new File(context.getFilesDir, BUSYBOX_FILENAME_EXTERN)
  }

  private def copyBusyBoxIfNotYetDone(context: Context):Boolean = {
    val appContext: Context = context.getApplicationContext
    if (!checkBusyBox(appContext)) {
      copyBusyBox(appContext)
      true
    } else false
  }

  /**
   * Copy BusyBox executable to the private data folder.
   *
   * @param context
   */
  private def copyBusyBox(context: Context) = {
      val busybox: File = getInternalBusybox(context)
      if (busybox.exists) {
//        require(busybox.delete, "Impossible to delete existing busybox")
      }
      val fos: FileOutputStream = context.openFileOutput(BUSYBOX_FILENAME_EXTERN, Context.MODE_PRIVATE)
      val is: InputStream = context.getAssets.open(BUSYBOX_FILENAME_INTERN)

      val bytes = new Array[Byte](1024) //1024 bytes - Buffer size
      Iterator
        .continually (is.read(bytes))
        .takeWhile (_ != -1 )
        .foreach (read=>fos.write(bytes,0,read))
        fos.close()

      require(busybox.length() > 0l, "Failed to copy busybox to this destination: " + busybox.getAbsolutePath)

      require(busybox.setReadable(false) && busybox.setWritable(false) && busybox.setExecutable(true, true), "Failed to change properties of this file: " + busybox.getAbsolutePath)
  }

  /**
   * Check if BusyBox application has been copied
   * in the private data folder of the application.
   *
   * @param context
   * @return true if already copied.
   */
  private def checkBusyBox(context: Context): Boolean = {
    val busybox: File = getInternalBusybox(context)
    busybox.exists && busybox.length == BUSYBOX_SIZE && busybox.canExecute
  }
}
